swig/cpp/tests/changes.cpp

```diff
@@ -0,0 +1,219 @@
+#include <iostream>
+#include <memory>
+#include <cassert>
+#include <cstring>
+#include <unistd.h>
+
+#include "Session.h"
+
+#define MAX_LEN 100
+
+using namespace std;
+
+const char *module_name = "swig-test";
+const int LOW_BOUND = 10;
+const int HIGH_BOUND = 20;
+const char *xpath_if_fmt = "/swig-test:cpp-operations/test-get[name='%s']/%s";
+
+char *
+get_test_name(int i)
+{
+    char buf[MAX_LEN] = {0};
+    sprintf(buf, "test-cpp-%d", i);
+    return strdup(buf);
+}
+
+const char *
+get_xpath(char *test_name, char *node_name)
+{
+    char buf[MAX_LEN];
+    sprintf(buf, xpath_if_fmt, test_name, node_name);
+    return strdup(buf);
+}
+
+class NOPCb: public Callback {
+public:
+    int module_change(S_Session sess, const char *module_name, sr_notif_event_t event, void *private_ctx)
+        {
+            return SR_ERR_OK;
+        }
+};
+
+void init_test(S_Session sess)
+{
+    S_Subscribe subs(new Subscribe(sess));
+    S_Callback cb(new NOPCb());
+
+    subs->module_change_subscribe(module_name, cb, NULL, 0, SR_SUBSCR_DEFAULT | SR_SUBSCR_APPLY_ONLY);
+
+    for (int i = LOW_BOUND; i < HIGH_BOUND; i++) {
+        const char *xpath = (char*)get_xpath(get_test_name(i), (char *) "number");
+        S_Val vset(new Val((int32_t)i, SR_INT32_T));
+        sess->set_item(xpath, vset);
+    }
+
+    sess->commit();
+    sess->copy_config(module_name, SR_DS_RUNNING, SR_DS_STARTUP);
+    usleep(100);
+    subs->unsubscribe();
+    usleep(1000);
+}
+
+void
+clean_test(S_Session sess)
+{
+    char xpath[100] = {0};
+    sprintf(xpath, "/%s:*", module_name);
+
+    sess->delete_item(xpath);
+    sess->commit();
+    // sess.copy_config(module_name, sr_datastore_t.SR_DS_RUNNING, sr_datastore_t.SR_DS_STARTUP);
+    usleep(100);
+}
+
+class DeleteCb: public Callback {
+public:
+    int module_change(S_Session sess, const char *module_name, sr_notif_event_t event, void *private_ctx)
+        {
+            char change_path[MAX_LEN];
+
+            try {
+                snprintf(change_path, MAX_LEN, "/%s:*", module_name);
+                auto it = sess->get_changes_iter(&change_path[0]);
+
+                while (auto change = sess->get_change_next(it)) {
+                    assert(SR_OP_DELETED == change->oper());
+                }
+            } catch (const std::exception& e) {
+                cout << e.what() << endl;
```

Comment:
This silently eats exceptions which is probably not something that you want to do in a unit test.

If you want to ignore an exception which is always thrown, add an assert that checks that it was thrown. If it doesn't really throw, then do not catch these exceptions and let the code crash -- it's a unit test. Crashing on unhandled exceptions is good.